### PR TITLE
Format.pp_print_custom_break, a more general break hint

### PR DESCRIPTION
**Pull request by keleshev**
`Format.pp_print_custom_break` is a more general break hint that can emit non-whitespace characters, for example, an additional trailing semicolon when a list of items is printed vertically.

# Motivation

Sometimes when formatting code or data structures you might want to print different non-whitespace characters depending on the layout being horizontal or vertical. For example, you might want to add a trailing semicolon when printing an OCaml list vertically:

```ocaml
(* Horizontal *)
[foo; baz; bar]

(* Vertical *)
[
  foo;
  bar;
  baz; (* Note the additional semicolon *)
]
```

Another example use-case is when a separator in a language is optional. For example, in many languages semicolons are optional for separating statements if each statement is located on a separate line. So you might want to remove semicolons when printing a bunch of statements vertically:

```
# Horizontal
{foo; baz; bar}

# Vertical
{
  foo
  baz
  bar
}
```

Both examples can be implemented using the new `Format.pp_print_custom_break`. See the added test file for the implementations.

# Implementation notes

The new break hint is implemented by introducing a new pretty-printing token `Pp_custom_break`. It is handled similarly to the existing `Pp_break`, with some modifications. If desirable, the `Pp_break` can be removed and `pp_print_break` can be re-implemented in terms of `Pp_custom_break`.

### CODE REVIEW

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: Perhaps add a new line with comment for consistency with surrounding code

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: It looks like this could be implemented in terms of `break_new_line`, or both could be refactored slightly to reduce repetition.

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml:411]: Also here, if the change is approved it would be better to refactor to avoid repetition with the `Pp_break` case. Or just remove or commandeer `Pp_break`.

**Review by jberdine — COMMENTED** [review]
The code looks good to me, though I have not tested it much. Best to refactor to remove repetition if the change is approved.

As a vote for utility, note that I made a similar addition [here](https://github.com/ocaml-ppx/ocamlformat/commit/caff7c4b98f33afb80c15c3fca128391dbcbf8f9) using a similar implementation, which is used extensively in ocamlformat.

**Comment by Drup** [comment]
I think the addition is useful and the API looks good. I'm not completely convinced by the labels, but that's a minor point. I agree with @jberdine's remarks: you should definitely express everything that can be in term of this new more general primitive.

@jberdine This is slightly offtopic, but is there somewhere a list of your local improvements for format in ocamlformat's fork ? Integrating practical improvements to Format in the mainline version would definitely interest me.

**Comment by jberdine** [comment]
(@drup, there is #1229, but it is slightly behind [ocamlformat_suppport.Format_](https://github.com/ocaml-ppx/ocamlformat/tree/support) and is embarrassingly in need of rebasing.)

**Comment by Drup** [comment]
@jberdine Oh, I missed that one. It seems to me that the current version superseeds your PR, except for "fits_or_breaks". Is that correct ?

**Comment by jberdine** [comment]
I'm not sure that `pp_print_string_if_newline` and `pp_print_or_newline` are expressible in terms of this PR. They are more similar to `pp_print_if_newline` in that they test if the current line has been broken, rather than being a break hint themselves. I think `pp_print_pre_break` is subsumed by this PR, but have not tested (I have been surprised by corners of `Format` behavior before).

**Comment by keleshev** [comment]
@jberdine I've added a test that shows that `pp_print_custom_break` covers the `pp_print_string_if_newline` use-case from #1229. I hope to make sure that this PR is useful for `ocamlformat`, and not something that you will need work around!

I'll re-implement `Pp_break` (and related code) in terms of `Pp_custom_break`. By that time `Pp_custom_break` will probably be simply called `Pp_break`.

**Comment by jberdine** [comment]
Note that the main point of the "test if the current line has been broken" behavior of e.g. `pp_print_string_if_newline` is to enable the flexibility of separating the break hint and the conditionally formatted string, in particular in cases where there is a `close_box` between them.

**Comment by pierreweis** [comment]
It is a good idea to generalize break hints. However, the philosophy of the
library is to provide a high level API to the programmer and thus to provide
formatting annotations directly in the format string arguments of *printf high
level functions. Could you explain how do you plan to provide this new breaks
in format strings ?

Another semantic problem is the behavior of the new break within compacting
boxes such as hov or structural hov boxes: the given motivations seem to
address the simple cases where the layout is statically known to be either
horizontal or vertical as within a hbox, a vbox or a hvbox. But what happens
in a compacting box ?

For instance, I generally want to print long lists neither horizontally nor
vertically but in a compacting box in order to get a more compact display like
this one:
[
  1; 2; 3; 4; 5; 6;
  7; 8; 9; 10; 11;
  12; 13; 14
]
To obtain this effect with the actual break hints, you need to have a
particular case for the last element of the list to omit the last semicolon.
Can I suppress this particular case and use the same new break hints for each
element ?

For the second motivations, would we get the following display:
{
  foo; bar; 
  geeeeeee
  buz
}
when printing the list of statements [foo; bar; geeeeeee; buz] in an hovbox ?
Or are the new break hints only used for the last item of the list of items ?

**Comment by keleshev** [comment]
@jberdine could you explain a bit more? Perhaps, an example? 

Edit: and most importantly, would there be cases in `ocamlformat` where `pp_print_string_if_newline` would do the job and `pp_print_cusom_break` would *not* be enough?

**Comment by keleshev** [comment]
> However, the philosophy of the
> library is to provide a high level API to the programmer and thus to provide
> formatting annotations directly in the format string arguments of *printf high
> level functions. Could you explain how do you plan to provide this new breaks
> in format strings ?

Currently, there is no special syntax for this break hint. But it works well with `%t` (from the test):

```ocaml
        fprintf ppf "[@;<0 2>@[<%s>%a@]%t]" box_type
          (list ~pp_sep (format_array box_type)) items
          (pp_print_custom_break ~no_break:"" ~yes_break:(";", ""))
```

Another way would be to provide a special syntax, for example `{|@;<"" ";" "">|}`. I'm interested to hear opinions about that.

In the three current tests/examples, the new break is used as 1) last break, 2) separator, 3) first break. I don't see it as a problem.

I will try to write more tests to see how the new breaks works with compacting boxes.

**Comment by keleshev** [comment]
In the last commit I added tests for compacting boxes. (I will clean up the history once we are happy with how the code looks. I think the "emergent" commit history is valuable during review, but not after it is merged.)

I am quite happy with how the custom break works with compacting boxes (see tests):

```
[
  Foo;
  Baz;
  Bar;
  Qux;
  Quux;
  Quuz;
  Corge;
  Grault;
  Garply;
]
[
  Foo; Baz; Bar; Qux;
  Quux; Quuz; Corge;
  Grault; Garply;
]
[Foo; Baz; Bar; Qux; Quux; Quuz; Corge; Grault; Garply]

{
  Foo
  Baz
  Bar
  Qux
  Quux
  Quuz
  Corge
  Grault
  Garply
}
{
  Foo; Baz; Bar; Qux
  Quux; Quuz; Corge
  Grault; Garply
}
{Foo; Baz; Bar; Qux; Quux; Quuz; Corge; Grault; Garply}
```

**Comment by Drup** [comment]
Nice work with the tests!

Regarding the syntax, `@;<"" ";" "">` is going to be a bit painful, due to string escaping (also, format syntax generally doesn't use `"`), but the underlying idea sounds good to me, we just need to find an appropriate separator. Until now, format uses spaces as separator, which is obviously not appropriate here.

**Comment by keleshev** [comment]
I think that this custom break is so rare in practice, that I would not give it a special syntax.

**Comment by keleshev** [comment]
Regarding implementing `pp_print_break` in terms of `pp_print_custom_break`. I have done this in a separate branch:

https://github.com/keleshev/ocaml/commit/cbdcb9122e1bb53fb07a4dfa1d0b548ebd1d33c5

There are some issues with that. `pp_print_break` uses `state.pp_out_spaces` callaback for same-line offset. If the offset is generalized to any string, then `state.pp_out_string` should be used instead. However, if so, then `state.pp_out_spaces` becomes an unused callback for all formatters! This is one breaking change.

Similar with new-line offset. In `pp_print_break` it uses `state.pp_out_indent` for printing all of the indentation + offset. However, if implemented with `pp_print_custom_break` the offset must be printed with `state.pp_out_string`. That's another breaking change.

Both could be fixed by peeking into the offset to check if it consists only of spaces, but that feels like a dirty hack to me. This also can influence performance, if we need to iterate over each char in offset each time.

Given the above, I lean towards the current implementation, which does *not* re-implement `pp_print_break` in terms of `pp_print_custom_break`.

**Comment by Drup** [comment]
That's an interesting question. While it's possible in theory to have `pp_out_space n` not being equivalent to `pp_out_string (String.create n ' ')`, I'm not sure anyone is doing that in the wild. In this case, doing that particular breaking change would be fine (given an appropriate removal path which neds to be discussed). `pp_out_indent` is still there and it's clearly the most useful of the two. I agree that analyzing the string does not sound like a good idea (if people want to do that, they can do it themselves in `pp_out_string`).

I think we need a bit of a survey of the ecosystem to know if we can make that simplification. 

@pierreweis Do you have any remarks on using `pp_out_string` instead of `pp_out_space` for (generalized) breaks ?

**Comment by keleshev** [comment]
> I think we need a bit of a survey of the ecosystem to know if we can make that simplification.

How is something like that usually done?

**Comment by keleshev** [comment]
FYI, `state.out_indent` was added in #615 in 4.06.0

**Comment by pierreweis** [comment]
I think we should not work with examples but work with semantics :)

1) We cannot suppress current break annotations from within format
strings. There is a lot of code using pretty-printing annotations such as
'@ ', '@,' '@< n m >' and we cannot suppress that.

2) The semantics of break hints depends of
   - the current enclosing box kind (b, h, v, hv, hov),
   - the current interpretation of indentation, spaces and line breaking
     which is local to each formatter,
   - the values of break hint parameters n and m.

Corollary: a break hint cannot be expressed in terms of newline charcter
emission and ASCII 32 character output for spaces and indentation. In some
sense, a break hint simply emits events to the underlying formatter, namely
line splitting signaling, space emission and indentation modification. A
formatter may deal with those signals as specified by the user that designed
the formatter. The definition of symbolic formatters explains and exemplifies
those concepts: the symbolic output items for spaces, newline and indentation
are purely symbolic and does not refer to any strings of characters.

Evidently, the simpler formatter provided by the Format module uses
characters and strings for all signals: newline is simply output_string
"\n", spaces n is output_string (String.make n ' ') and indent n is spaces n.

But, more sophisticated formatters are common, for instance
 - newline is simply output_string "<BR>",
 - spaces n is n times output_string "&nbsp;"
 - indent n is still spaces n

In short, you cannot oversimplify signals to convert them to output strings
because you don't know which strings are appropriate to the application at
hand.

3) Also note that the second parameter of a break hint (the one that
   specifies the modification of current indentation when the break lead to
   line splitting) can be negative. It means that a break hint may reduce the
   current indentation!
   Compare positive indentation modification:
     # printf "@[<v 2>%d@;<1 0>%d@]" 1 2;;
     1
       2- : unit = ()
   with negative indentation modification:
     # printf "@[<v 2>%d@;<1 -2>%d@]" 1 2;;
     1
     2- : unit = ()

   I use it a lot in pretty-printing routines to print something at the
   indentation level of the enclosing box.

In short, more powerful break hints may be a desirable thing, but the new
feature should not break the existing generality of the current
pretty-printer.

So let's describe the semantics of existing break hints, then a proposition
of an enhanced semantics for extended break hints.

Semantics of actual break hint '@;<n m>' is twofold:
 - non splitting semantics is: emit a (space n) signal
 - splitting semantics is in turn:
   * emit a newline signal,
   * emit an indentation modification of current indentation by adding m

For simplicity of exposition let's simply express the additional strings in
extended break hints specifications as %s indications in the format string.
Now, instead of simple break hint @;<n m>, an extended break hint could be in
its full generality

  @;<%s n %s, %s m %s>
  (for visibility an additional comma separates the splitting and
  non-splitting part of the specification)

Semantics of extended break hint @;<%s n %s, %s m %s> would be (let's name
the various strings s1 s2 s3 s4):
 - non splitting semantics is in turn:
   * emit string s1,
   * emit a (space n) signal,
   * emit string s2
 - splitting semantics is in turn:
   * emit string s3,
   * emit a newline signal,
   * emit an indentation modification of current indentation by adding m,
   * emit string s4

My intuition is that all strings s1, s2, s3, and s4 would count in the
calculation of the current line length (as opposed to semantic tag
marking). This has to be carefully thought out to choose between
non-splitting and splitting semantics for the break hint according to the
length of string s1 and s2.

If we assume that extra strings in extended break hints may be omitted we
clearly recover the current break hints and their semantics.

This design would preserve existing break hints for backward compatibility
and implement your current proposition.

**Comment by keleshev** [comment]
@pierreweis this sounds very interesting! I will try out this design. Good point about the negative offset—that's yet another reason why the current proposal is not a generalization of the existing break hint.

**Comment by jberdine** [comment]
> @jberdine could you explain a bit more? Perhaps, an example?

I don't have a canned example, but one use is `if_newline` 
[here](https://github.com/ocaml-ppx/ocamlformat/blob/72fd899628cf0787afaaaefac882948506a70380/src/Fmt_ast.ml#L1546) 
and the usage can be seen by following how that value flows into 
`fmt_pattern`.

**Comment by jberdine** [comment]
> For simplicity of exposition let's simply express the additional 
> strings in
> extended break hints specifications as %s indications in the format 
> string.
> Now, instead of simple break hint @;<n m>, an extended break hint 
> could be in
> its full generality
>
>   @;<%s n %s, %s m %s>
>   (for visibility an additional comma separates the splitting and
>   non-splitting part of the specification)
>
> Semantics of extended break hint @;<%s n %s, %s m %s> would be (let's 
> name
> the various strings s1 s2 s3 s4):
>  - non splitting semantics is in turn:
>    * emit string s1,
>    * emit a (space n) signal,
>    * emit string s2
>  - splitting semantics is in turn:
>    * emit string s3,
>    * emit a newline signal,
>    * emit an indentation modification of current indentation by adding 
> m,
>    * emit string s4

I think that this is a nice generalization. Just as a data point 
supporting the generality: the `pp_print_pre_break` I mentioned above is 
as far as I understand also a specialization: where all but s3 are 
omitted.

**Comment by keleshev** [comment]
I have implemented @pierreweis's design and am ready for review.

**Comment by rbonichon** [comment]
@keleshev You have indeed handled the printing aspects of @pierreweis 's suggestion. 
My reading is that he also mentions generalizing the format string formatting command, like `@;<%s n %s, %s m %s>` which could be specified as the current break hints. 

In a way this respects a kind of (implicit) symmetry where you can use direct formatting functions or formatted format strings for you pretty-printing specifications (for example: `pp_print_cut` and `@,`).

I do not see that your proposal handles that aspect --- and this may actually make your pull request more involved, touching GADT's and all.
Would you be keen on adding that as well ?

**Comment by alainfrisch** [comment]
My 2 cents: format strings are good as long as format remain relatively simple, but I don't see a strong need for creating complex syntax for advanced concepts.  They are better served (in term or code readability, tooling, etc) by normal functions.

**Comment by jberdine** [comment]
I agree with @alainfrisch. I think format strings are great for short or simple uses, but I have found that they become hard to read, and are less flexible, for more complicated uses. And for code that primarily uses format strings, there a `%a` specifier can be used for the added custom break hints. If adding format strings for this is not too involved, it would be good to do. But I would hate to see this PR get blocked on it.

(I plan to look at this PR in detail, but probably won't have time until early next week.)

**Comment by keleshev** [comment]
It is intentional that there's no special syntax for the new break. I think it is too rare to warrant a special syntax, and it works well enough with `%t`. Example:

```ocaml
        fprintf ppf "[@;<0 2>@[<%s>%a@]%t]" box_type
          (list ~pp_sep (format_array box_type)) items
          (pp_print_custom_break ~no_break:("", 0, "") ~yes_break:(";", 0, ""))
```

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: I don't really want to bike-shed the names much, but just note that given the name `Pp_fits` used for boxes that do not break, it might be a nice analogy to use `fits` here for `no_break`, and then `breaks` for `yes_break` as its complement.

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: I suspect that this should not be unconditionally set to `false`, for instance when used on an empty `before` or `after` string.

I'd consider not clearing this flag for `Pp_text` of an empty string to be a bug fix wrt the current behavior.

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: My understanding is that `length` should be the number of characters consumed in the non-breaking case. So shouldn't `width` also include the lengths of the `no_break` strings?

File: Changes
Comment (jberdine) [inline Changes]: I think it would be appropriate to also note the design input by Pierre Weis.

File: stdlib/format.mli
Comment (jberdine) [inline stdlib/format.mli:300]: Do others think that it would be helpful to explicitly note that `out_spaces` is used to emit the spaces, to head off user confusion about why `(s1, n, s2)` isn't just one string (with some space characters in the middle)?

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: I think you need to add the length of the first string of yes_break to size here.

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: And here.

File: testsuite/tests/lib-format/pp_print_custom_break.ml
Comment (jberdine) [inline testsuite/tests/lib-format/pp_print_custom_break.ml]: It would be good to test hv and hov boxes as well, especially as they involve more complicated breaking calculations.

File: Changes
Comment (gasche) [reply Changes]: (If it helps inspiration, on similar occasions I have used the equivalent of "Vladimir Keleshev and Pierre Weis". There is an asymmetry that isn't fully rendered, but I'm sure @keleshev will make many other nice contributions so that being fully precise about each one matters less.)

File: stdlib/format.ml
Comment (gasche) [reply stdlib/format.ml]: Could we get a regression test against this?

**Review by gasche — APPROVED** [review]
In order to limit uncertainty about the work being poured into this PR, I'm approving it (but I have not reviewed it myself) on the basis of:

- the choice of API suggested by @pierreweis  (thanks!)
- the thorough implementation review by @jberdine (thanks!)

This should not be merged immediately, in particular there remain comments by @jberdine to address -- then rebase the PR, check the CI, etc. @keleshev, please say explicitly when you think that the PR has reached a mergeable state (including a cleaned-up commit history). And thanks!

**Comment by gasche** [comment]
Gentle ping. There are conflicts in the Changes entry now, so this would need a rebase. @keleshev, are you interested in integrating @jberdine's feedback?

**Comment by keleshev** [comment]
@gasche yes, I'm planning to address all the excellent @jberdine suggestions. I hope that I will have time to do this during November.

**Comment by keleshev** [comment]
@jberdine I've addressed your suggestions. Can you have another look?

**Comment by jberdine** [comment]
Thanks, I will take a look, hopefully by Wed next week.

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: Any reason to prefer testing `text <> ""` over `size > 0`?

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: I'm not sure of the standard style, but I would add parens around the tuple to avoid relying on the relative precedence of `,` vs `as`.

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: *shrug*

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: Same comment re precedence of `,` vs `as`.

File: stdlib/format.ml
Comment (jberdine) [inline stdlib/format.ml]: With the possibility of users overriding the default `pp_out_string` function, there might be a difference between calling `pp_out_string` on an empty string vs not calling `pp_out_string` at all. Perhaps `format_pp_text` should only be called when `s <> ""`? Then the `text <> ""` test in `format_pp_text` could be removed, preserving the possibly unintentional current behavior of clearing `pp_is_new_line` when printing an empty string.

**Review by jberdine — COMMENTED** [review]
Looks good. I have only one suggestion, and a few very minor comments.

**Comment by keleshev** [comment]
@jberdine how about now?

**Comment by jberdine** [comment]
Thanks! This looks good to me.

**Comment by keleshev** [comment]
@gasche this is ready to be merged, I think.

**Comment by gasche** [comment]
Merged! Thanks everyone for your work.

**Pull request merged**
